go/test/endtoend/tabletmanager/throttler/throttler_test.go

```diff
@@ -179,34 +167,48 @@ func warmUpHeartbeat(t *testing.T) (respStatus int) {
 func TestThrottlerAfterMetricsCollected(t *testing.T) {
 	defer cluster.PanicHandler(t)
 
-	time.Sleep(throttlerInitWait)
-	// By this time metrics will have been collected. We expect no lag, and something like:
-	// {"StatusCode":200,"Value":0.282278,"Threshold":1,"Message":""}
-	//
-	respStatus := warmUpHeartbeat(t)
-	assert.NotEqual(t, http.StatusOK, respStatus)
-	time.Sleep(time.Second)
-	{
+	// We run with on-demand heartbeats. Immediately as the tablet manager opens, it sends a one-time
+	// request for heartbeats, which means the throttler is able to collect initial "good" data.
+	// After a few seconds, the heartbeat lease terminates. We wait for that.
+	// {"StatusCode":429,"Value":4.864921,"Threshold":1,"Message":"Threshold exceeded"}
+	t.Run("expect push back once initial heartbeat lease terminates", func(t *testing.T) {
+		time.Sleep(onDemandHeartbeatDuration + applyConfigWait)
+		resp, err := throttleCheck(primaryTablet)
```

Comment:
I see what you mean, but there is an inherent problem to this approach: we use on-demand heartbeats. When we _check_ the throttler, this kicks and renews a heartbeat lease. Now, we want to validate that if we _don't do anything_ for a predefined amount of time, then the heartbeats stop, leading to the throttler returning an error.

Now, your request is that instead of sleeping the amount of time, we continuously poll until we reach the desired response code. This is something that works well in most tests to mitigate the CI's crazy timings. However, this will not work here, because the mere fact we make a _check_ means we affect the behavior of the throttler. The check will kick a new heartbeats lease, where we really want it to _not renew_.

So the checks defeat the purpose.

We _could_ possibly run a check with a "please-do-not-renew-lease" flag. Let me see about that.

Reply:
OK this is done! Added a new check flag which disables hearbeat lease renewal; all tests are now wait-based unless strictly required not to be.